signals/signal_exporter.py

```diff
@@ -0,0 +1,383 @@
+                if "ras" in data:
+                    to_check.remove(data["tool_id"])
+                    if data["ras"] != 1:
+                        print(
+                            f"WARNING: Tool '{data['tool_id']}' returned bad response for event '{event}', ras: {data['ras']}"
+                        )
+                        result_box[0] = 1
+            if not to_check:
+                listener.stop()
+
+        subscriber.subscribe(**{"event-signal-response": _sub_handler})
+        listener = subscriber.run_in_thread()
+        return listener, result_box
+
+    def _valid_str_list(self, names: list) -> bool:
+        """
+        Return true if input is a non-empty list of strings. Otherwise
+        return false.
+        """
+        return (
+            bool(names)
+            and isinstance(names, list)
+            and all(isinstance(event, str) for event in names)
+        )
+
+    def publish_signal(
+        self, event: str, sample: int = -1, tag: str = None, metadata: Dict = None
+    ) -> int:
+        """
+        Publish a legal event signal. Includes additional options to specify sample_no,
+        a tag, and any other additional metadata. Will then wait for responses from
+        subscribed responders (if any). Returns one of the below result codes based on 
+        signal publish/response success.
+
+        RESULT CODES:
+        0 = ALL SUBS RESPONDED WELL
+        1 = SOME SUBS RESPONDED BADLY
+        2 = NOT ALL SUBS RESPONDED
+        3 = ILLEGAL EVENT NAME PASSED IN
+        4 = INITIALIZATION SIGNAL ATTEMPTED, BAD
+        5 = SHUTDOWN SIGNAL ATTEMPTED, BAD
+        """
+        skip_check = False
+        if not self.init_listener or not self.init_listener.is_alive():
+            print(
+                "WARNING: Exporter is not initialized, not accepting subscribers and no event checking"
+            )
+            skip_check = True
+
+        if event == "initialization":
+            print(
+                "ERROR: Please use the 'initialize()' method for publishing initialization signals"
+            )
+            return 4
+
+        if event == "shutdown":
+            print("ERROR: Please use the 'shutdown()' method for shutdown signals")
+            return 5
+
+        if not skip_check and not event in self.legal_events:
+            print(f"Event {self.event} not one of legal events: {self.legal_events}")
+            return 3
+
+        sig = self._sig_builder(event=event, sample=sample, tag=tag, metadata=metadata)
+        sub_check, result_box = self._check_subs(event)
+
+        self.redis.publish(channel="event-signal-pubsub", message=sig.to_json_str())
+
+        counter = 0
+        while sub_check and sub_check.is_alive():
+            time.sleep(0.1)
+            counter += 1
+            if counter >= 200:
+                print("Timeout after waiting 20 seconds for sub response")
+                return 2
```

Comment:
Just noticed this: do we need to stop the ``sub_check`` listener here? I'm concerned that it will keep working in the background even if we have encountered a timeout.

Reply:
We definitely do lol, nice catch